File: scala/stack/Stack.scala
import scala.util.control.Breaks._
class Stack[T](val size: Int = 10){
  private var top = -1
  private var stack: List[T] = Nil
  def isEmpty = top == -1
  def isFull = top+1 == size
  def push(e: T) = {
    if(isFull) println("StackOverflow") 
    else {
      stack = stack :+ e
      top += 1
    }
  }
  def peek = stack.lastOption.getOrElse(None)
  def pop = {
    if(top == -1) println("Stack Empty")
    else {
      val e = stack.last
      stack = stack.slice(0, top)
      top -= 1
      e
    }
  }
}


object StackTest extends App {
  val stack =  new Stack[Int]
  println(s"Stack Empty: ${stack.isEmpty}")
  println("Pusing elemnts to Stack ...")
  breakable {
    for(x <- 1 to 20) {
      if(stack.isFull) break
      stack.push(x)
    }
  }
  println(s"stack is Full: ${stack.isFull}")
  println("Popping OUT all elements")
  breakable {
    while(true) {
      if(stack.isEmpty) break
      println(stack.pop)
    }
  }
  stack.pop
}

File: python/stack/stack.py
class Stack:
  def __init__(self, size=10):
    self.stack = [None]*size
    self.size = 10
    self.top = -1

  def push(self, e):
    if self.isFull():
      print("StackOverflow")
    else:
      self.top += 1
      self.stack[self.top] = e

  def pop(self):
    if self.isEmpty():
       print("Stack Empty")
       return
    self.top -= 1
    return self.stack[self.top+1]

  def peek(self):
    return self.stack[self.top]

  def isEmpty(self):
      return (False, True)[self.top == -1]

  def isFull(self):
    return (False , True)[self.size == self.top+1]

stack = Stack()
print("Stack empty: {}".format(stack.isEmpty()))
for x in range(1,11):
  stack.push(x)
print("Peeking Stack: {}".format(stack.peek()))
print("Stack full: {}".format(stack.isFull()))

print("Stack Elements")
for x in range(10):
    print(stack.pop())
stack.pop()
